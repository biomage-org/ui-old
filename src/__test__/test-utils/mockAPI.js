import _ from 'lodash';

// Default platform data
import cellSetsData from '__test__/data/cell_sets.json';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockDemoExperiments from '__test__/test-utils/mockData/mockDemoExperiments.json';

// A ticket has been created to address this : https://biomage.atlassian.net/browse/BIOMAGE-1553
import {
  responseData,
} from '__test__/test-utils/mockData';

const promiseResponse = (
  response,
  options = {},
) => Promise.resolve(new Response(response, options));

const statusResponse = (code, body) => Promise.resolve({
  status: code,
  body: JSON.stringify(body),
});

const delayedResponse = (response, delay = 10000, options = {}) => new Promise((resolve) => {
  setTimeout(() => {
    resolve(new Response(response, options));
  }, delay);
});

const workerResponse = (body) => promiseResponse(body);

const generateDefaultMockAPIResponses = (experimentId) => ({
  [`experiments/${experimentId}`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === experimentId)),
  ),
  [`experiments/${experimentId}/processingConfig`]: () => promiseResponse(
    JSON.stringify(processingConfigData),
  ),
  [`experiments/${experimentId}/cellSets`]: () => promiseResponse(
    JSON.stringify(cellSetsData),
  ),
  [`experiments/${experimentId}/backendStatus`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
  experiments: () => promiseResponse(
    JSON.stringify(responseData.experiments),
  ),
  [`experiments/${experimentId}/samples`]: () => promiseResponse(
    JSON.stringify(responseData.samples[0]),
  ),
  '/v2/experiments/examples': () => promiseResponse(
    JSON.stringify(mockDemoExperiments),
  ),
});

const mockAPI = (apiMapping) => (req) => {
  const path = req.url;

  const key = _.find(
    Object.keys(apiMapping),
    (urlStub) => path.endsWith(urlStub),
  );

  if (!key) {
    return statusResponse({
      status: 404,
      body: `Path ${path} is undefined`,
    });
  }

  return apiMapping[key](req);
};

export default mockAPI;
export {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  workerResponse,
  delayedResponse,
};
