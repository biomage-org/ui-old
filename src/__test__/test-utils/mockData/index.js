import _ from 'lodash';
import generateMockSamples from './generateMockSamples';
import generateMockExperiments from './generateMockExperiments';
import generateMockProcessingConfig from './generateMockProcessingConfig';

// This file is a work in progress to generate mock data for the endpoints of the test.
// The generated data should be usable as response data to be sent via the mocked API in mockAPI.js
// The module should also expose the generated data that can be imported and used easily in tests
// A ticket has been created to give more thought into this https://biomage.atlassian.net/browse/BIOMAGE-1553

const responseData = {};

responseData.experiments = generateMockExperiments(2);

responseData.samples = [generateMockSamples(
  responseData.experiments[0],
  3,
)];

responseData.processingConfig = generateMockProcessingConfig(3);

// Add samples to first experiment
const samples = responseData.samples[0];
const sampleIds = _.map(samples, 'id');
responseData.experiments[0].samplesOrder = sampleIds;

const { experiments, processingConfig } = responseData;

export {
  experiments,
  samples,
  processingConfig,
  responseData,
};
