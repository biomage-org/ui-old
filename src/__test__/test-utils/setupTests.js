import '@testing-library/jest-dom/extend-expect';
import preloadAll from 'jest-next-dynamic';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('localforage');
jest.mock('utils/pushNotificationMessage');

beforeAll(async () => {
  // Add stuff that needs to run once, before all tests
  await preloadAll();
});

beforeEach(async () => {
  // Add stuff that needs to run before each test
});
