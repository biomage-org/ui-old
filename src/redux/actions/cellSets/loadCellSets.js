import fetchAPI from 'utils/http/fetchAPI';
import handleError from 'utils/http/handleError';
import {
  CELL_SETS_LOADED, CELL_SETS_LOADING, CELL_SETS_ERROR,
} from 'redux/actionTypes/cellSets';
import endUserMessages from 'utils/endUserMessages';

import config from 'config';
import { api } from 'utils/constants';

const loadCellSets = (experimentId, forceReload = false) => async (dispatch, getState) => {
  const {
    loading, error, updatingClustering,
  } = getState().cellSets;

  if (!forceReload && ((!loading && !error) || updatingClustering)) {
    return null;
  }

  if (error) {
    dispatch({
      type: CELL_SETS_LOADING,
    });
  }

  let url;

  if (config.currentApiVersion === api.V1) {
    url = `/v1/experiments/${experimentId}/cellSets`;
  } else if (config.currentApiVersion === api.V2) {
    url = `/v2/experiments/${experimentId}/cellSets`;
  }

  try {
    const data = await fetchAPI(url);

    dispatch({
      type: CELL_SETS_LOADED,
      payload: {
        experimentId,
        data: data.cellSets,
      },
    });
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_FETCHING_CELL_SETS);

    dispatch({
      type: CELL_SETS_ERROR,
      payload: { error: errorMessage },
    });
  }
};

export default loadCellSets;
