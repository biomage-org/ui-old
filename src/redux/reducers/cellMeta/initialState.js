const mitochondrialContent = {
  loading: true,
  error: false,
  data: [],
};

const doubletScores = {
  loading: true,
  error: false,
  data: [],
};

const numOfGenes = {
  loading: true,
  error: false,
  data: [],
};

const numOfUmis = {
  loading: true,
  error: false,
  data: [],
};

const initialState = {
  mitochondrialContent,
  doubletScores,
  numOfGenes,
  numOfUmis,
};

export default initialState;
