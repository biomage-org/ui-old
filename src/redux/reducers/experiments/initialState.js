const experimentTemplate = {
  name: null,
  description: 'Analysis description',
  id: null,
  createdAt: null,
  updatedAt: null,
  notifyByEmail: true,
  meta: { organism: null, type: '10x' },
  sampleIds: [],
  metadataKeys: [],
};

const initialState = {
  meta: {
    loading: false,
    error: false,
    saving: false,
    activeExperimentId: null,
  },
  ids: [],
};

const METADATA_DEFAULT_VALUE = 'N.A.';

export default initialState;
export { experimentTemplate, METADATA_DEFAULT_VALUE };
