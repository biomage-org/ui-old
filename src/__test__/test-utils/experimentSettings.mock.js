import initialState from 'redux/reducers/experimentSettings/initialState';

const sampleifiedConfig = (sampleIds, configToReplicate) => {
  const result = sampleIds.reduce(
    (acum, sampleId) => {
      // eslint-disable-next-line no-param-reassign
      acum[sampleId] = configToReplicate;
      return acum;
    },
    {},
  );

  return result;
};

const generateProcessingConfigMock = (sampleIds) => ({
  classifier: {
    enabled: true,
    prefiltred: false,
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        FDR: 0.1,
      },
    }),
  },
  cellSizeDistribution: {
    enabled: true,
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        minCellSize: 10800,
        binStep: 200,
      },
    }),
  },
  mitochondrialContent: {
    enabled: true,
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        method: 'absoluteThreshold',
        methodSettings: {
          absoluteThreshold: {
            maxFraction: 0.1,
            binStep: 200,
          },
        },
      },
    }),
  },
  numGenesVsNumUmis: {
    enabled: true,
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        regressionType: 'gam',
        regressionTypeSettings: {
          gam: {
            'p.level': 0.00009,
          },
        },
      },
    }),
  },
  doubletScores: {
    enabled: true,
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        probabilityThreshold: 0.2,
        binStep: 0.05,
      },
    }),
  },
  dataIntegration: {
    enabled: true,
    dataIntegration: {
      method: 'seuratv4',
      methodSettings: {
        seuratv4: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
        fastmnn: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
        unisample: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
      },
    },
    dimensionalityReduction: {
      method: 'rpca',
      numPCs: 30,
      excludeGeneCategories: [],
    },
  },
  configureEmbedding: {
    enabled: true,
    embeddingSettings: {
      method: 'umap',
      methodSettings: {
        umap: {
          minimumDistance: 0.1,
          distanceMetric: 'euclidean',
        },
        tsne: {
          perplexity: 30,
          learningRate: 200,
        },
      },
    },
    clusteringSettings: {
      method: 'louvain',
      methodSettings: {
        louvain: {
          resolution: 0.5,
        },
      },
    },
  },
});

const generateExperimentSettingsMock = (sampleIds) => {
  const mockedProcessingConfig = {
    ...initialState.processing,
    ...generateProcessingConfigMock(sampleIds),
  };

  return {
    ...initialState,
    processing: mockedProcessingConfig,
    originalProcessing: mockedProcessingConfig,
    info: {
      ...initialState.info,
      pipelineVersion: 1,
    },
  };
};

export default generateExperimentSettingsMock;
