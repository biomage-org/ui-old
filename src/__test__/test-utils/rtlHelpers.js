import { fireEvent, screen } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';
import { act } from 'react-dom/test-utils';

const selectOption = async (optionName, select) => {
  await userEvent.click(select);
  // userEvent click doesnt trigger the onChange in select
  fireEvent.click(screen.getByTitle(optionName));
};

// dropFilesIntoDropzone takes an input element and simulates droping files into it
// inputElement: this needs to be the actual input element not the dropzone itself, e.g.:
// * <input data-testid='drop-input'/> => screen.getByTestId('drop-input');
// files should be a list of files, e.g.:
// * new File(['content'], 'file.txt', { type: 'text/plain' });
const dropFilesIntoDropzone = async (inputElement, files) => {
  await act(async () => {
    fireEvent.drop(inputElement, {
      target: { files },
    });
  });
};

export {
  // eslint-disable-next-line import/prefer-default-export
  selectOption,
  dropFilesIntoDropzone,
};
