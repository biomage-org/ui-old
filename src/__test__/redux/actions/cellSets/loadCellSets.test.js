import _ from 'lodash';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';

import loadCellSets from 'redux/actions/cellSets/loadCellSets';
import initialState from 'redux/reducers/cellSets/initialState';

import '__test__/test-utils/setupTests';
import { CELL_SETS_ERROR, CELL_SETS_LOADED, CELL_SETS_LOADING } from 'redux/actionTypes/cellSets';

enableFetchMocks();
const mockStore = configureStore([thunk]);

describe('loadCellSets action', () => {
  const experimentId = '1234';

  const cellSets = [
    {
      key: 'louvain',
      name: 'louvain clusters',
      rootNode: true,
      type: 'cellSets',
      children: [
        {
          key: 'louvain-1',
          name: 'Cluster 1',
          rootNode: false,
          type: 'cellSets',
          color: '#ee8866',
          cellIds: [5, 18, 33, 55, 77],
        },
        {
          key: 'louvain-0',
          name: 'Cluster 0',
          rootNode: false,
          type: 'cellSets',
          color: '#000000',
          cellIds: [11, 23, 27, 48, 61],
        },
        {
          key: 'louvain-3',
          name: 'Cluster 3',
          rootNode: false,
          type: 'cellSets',
          color: '#ffaabb',
          cellIds: [68, 112, 122, 132, 189],
        },
        {
          key: 'louvain-2',
          name: 'Cluster 2',
          rootNode: false,
          type: 'cellSets',
          color: '#eedd88',
          cellIds: [3, 34, 42, 49, 71],
        },
      ],
    },
    {
      key: 'scratchpad',
      name: 'Custom cell sets',
      rootNode: true,
      children: [
        {
          key: '8297d513-6268-4ab0-a03e-e74a23ecec07',
          name: 'New Cluster',
          color: '#3957ff',
          type: 'cellSets',
          cellIds: [3103, 449, 4023, 2545, 930],
        },
        {
          key: '99a7746b-7f9d-4e54-8acf-53031e4ff023',
          name: 'New Cluster2',
          color: '#d3fe14',
          type: 'cellSets',
          cellIds: [6536, 3654, 1075, 1538, 193],
        },
      ],
      type: 'cellSets',
    },
    {
      key: 'sample',
      name: 'Samples',
      rootNode: true,
      children: [
        {
          key: 'b15fa0d6-74e8-49da-bcbb-aa07127960d0',
          name: 'Convalescent',
          color: '#77aadd',
          cellIds: [0, 1, 2, 4, 5, 7],
        },
      ],
      type: 'metadataCategorical',
    },
  ];

  const experimentSettings = {
    info: {
      sampleIds: [],
    },
  };

  beforeEach(() => {
    const response = new Response(JSON.stringify({ cellSets }));

    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockResolvedValueOnce(response);
  });

  it('Does not dispatch on normal operation', async () => {
    const store = mockStore({
      cellSets: {
        initialLoadPending: false, loading: false, error: false, updatingClustering: false,
      },
      experimentSettings,
    });

    await store.dispatch(loadCellSets(experimentId));
    expect(store.getActions().length).toEqual(0);
  });

  it('Dispatches on force reload', async () => {
    const store = mockStore({ cellSets: { loading: false, error: false }, experimentSettings });
    await store.dispatch(loadCellSets(experimentId, true));

    expect(_.map(store.getActions(), 'type')).toEqual([CELL_SETS_LOADING, CELL_SETS_LOADED]);
  });

  it('Dispatches on initial load pending false and error state', async () => {
    const store = mockStore({
      cellSets: { initialLoadPending: false, loading: false, error: true },
      experimentSettings,
    });

    await store.dispatch(loadCellSets(experimentId));

    expect(_.map(store.getActions(), 'type')).toEqual([CELL_SETS_LOADING, CELL_SETS_LOADED]);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('Dispatches a loaded action when run with the initial state.', async () => {
    const store = mockStore({ cellSets: initialState, experimentSettings });
    await store.dispatch(loadCellSets(experimentId));

    expect(_.map(store.getActions(), 'type')).toEqual([CELL_SETS_LOADING, CELL_SETS_LOADED]);

    expect(store.getActions()).toMatchSnapshot();
  });

  it('Dispatches an error condition if fetch fails', async () => {
    const store = mockStore({ cellSets: initialState, experimentSettings });

    fetchMock.resetMocks();
    fetchMock.mockReject(new Error('some weird error that happened'));

    await store.dispatch(loadCellSets(experimentId));

    expect(_.map(store.getActions(), 'type')).toEqual([CELL_SETS_LOADING, CELL_SETS_ERROR]);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('Uses V2 URL when using API version V2', async () => {
    const store = mockStore({ cellSets: initialState, experimentSettings });
    await store.dispatch(loadCellSets(experimentId));

    const fetchUrl = fetchMock.mock.calls[0][0];

    expect(fetchUrl).toEqual('http://localhost:3000/v2/experiments/1234/cellSets');
  });
});
