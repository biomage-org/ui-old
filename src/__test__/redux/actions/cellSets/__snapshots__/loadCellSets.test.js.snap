// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadCellSets action Dispatches a loaded action when run with the initial state. 1`] = `
[
  {
    "type": "cellSets/loading",
  },
  {
    "payload": {
      "data": [
        {
          "children": [
            {
              "cellIds": [
                5,
                18,
                33,
                55,
                77,
              ],
              "color": "#ee8866",
              "key": "louvain-1",
              "name": "Cluster 1",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                11,
                23,
                27,
                48,
                61,
              ],
              "color": "#000000",
              "key": "louvain-0",
              "name": "Cluster 0",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                68,
                112,
                122,
                132,
                189,
              ],
              "color": "#ffaabb",
              "key": "louvain-3",
              "name": "Cluster 3",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                3,
                34,
                42,
                49,
                71,
              ],
              "color": "#eedd88",
              "key": "louvain-2",
              "name": "Cluster 2",
              "rootNode": false,
              "type": "cellSets",
            },
          ],
          "key": "louvain",
          "name": "louvain clusters",
          "rootNode": true,
          "type": "cellSets",
        },
        {
          "children": [
            {
              "cellIds": [
                3103,
                449,
                4023,
                2545,
                930,
              ],
              "color": "#3957ff",
              "key": "8297d513-6268-4ab0-a03e-e74a23ecec07",
              "name": "New Cluster",
              "type": "cellSets",
            },
            {
              "cellIds": [
                6536,
                3654,
                1075,
                1538,
                193,
              ],
              "color": "#d3fe14",
              "key": "99a7746b-7f9d-4e54-8acf-53031e4ff023",
              "name": "New Cluster2",
              "type": "cellSets",
            },
          ],
          "key": "scratchpad",
          "name": "Custom cell sets",
          "rootNode": true,
          "type": "cellSets",
        },
        {
          "children": [
            {
              "cellIds": [
                0,
                1,
                2,
                4,
                5,
                7,
              ],
              "color": "#77aadd",
              "key": "b15fa0d6-74e8-49da-bcbb-aa07127960d0",
              "name": "Convalescent",
            },
          ],
          "key": "sample",
          "name": "Samples",
          "rootNode": true,
          "type": "metadataCategorical",
        },
      ],
      "experimentId": "1234",
    },
    "type": "cellSets/loaded",
  },
]
`;

exports[`loadCellSets action Dispatches an error condition if fetch fails 1`] = `
[
  {
    "type": "cellSets/loading",
  },
  {
    "payload": {
      "error": "We couldn't get the list of cell sets.",
    },
    "type": "cellSets/error",
  },
]
`;

exports[`loadCellSets action Dispatches on initial load pending false and error state 1`] = `
[
  {
    "type": "cellSets/loading",
  },
  {
    "payload": {
      "data": [
        {
          "children": [
            {
              "cellIds": [
                5,
                18,
                33,
                55,
                77,
              ],
              "color": "#ee8866",
              "key": "louvain-1",
              "name": "Cluster 1",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                11,
                23,
                27,
                48,
                61,
              ],
              "color": "#000000",
              "key": "louvain-0",
              "name": "Cluster 0",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                68,
                112,
                122,
                132,
                189,
              ],
              "color": "#ffaabb",
              "key": "louvain-3",
              "name": "Cluster 3",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                3,
                34,
                42,
                49,
                71,
              ],
              "color": "#eedd88",
              "key": "louvain-2",
              "name": "Cluster 2",
              "rootNode": false,
              "type": "cellSets",
            },
          ],
          "key": "louvain",
          "name": "louvain clusters",
          "rootNode": true,
          "type": "cellSets",
        },
        {
          "children": [
            {
              "cellIds": [
                3103,
                449,
                4023,
                2545,
                930,
              ],
              "color": "#3957ff",
              "key": "8297d513-6268-4ab0-a03e-e74a23ecec07",
              "name": "New Cluster",
              "type": "cellSets",
            },
            {
              "cellIds": [
                6536,
                3654,
                1075,
                1538,
                193,
              ],
              "color": "#d3fe14",
              "key": "99a7746b-7f9d-4e54-8acf-53031e4ff023",
              "name": "New Cluster2",
              "type": "cellSets",
            },
          ],
          "key": "scratchpad",
          "name": "Custom cell sets",
          "rootNode": true,
          "type": "cellSets",
        },
        {
          "children": [
            {
              "cellIds": [
                0,
                1,
                2,
                4,
                5,
                7,
              ],
              "color": "#77aadd",
              "key": "b15fa0d6-74e8-49da-bcbb-aa07127960d0",
              "name": "Convalescent",
            },
          ],
          "key": "sample",
          "name": "Samples",
          "rootNode": true,
          "type": "metadataCategorical",
        },
      ],
      "experimentId": "1234",
    },
    "type": "cellSets/loaded",
  },
]
`;
